improver/ensemble_copula_coupling/utilities.py

```diff
@@ -289,3 +259,100 @@ def restore_non_percentile_dimensions(
+    used with the three long 1-dimensional arrays in a single step.
+
+    For a typical spatial grid the vectorisation applied along say x, but a loop
+    is used over y. This reduces the amount of memory used whilst providing a
+    speed up.
+
+    This method will become slower for increasingly large percentile coordinates
+    but provides a significant speed up for typical numbers of percentiles
+    (i.e. up to around 20).
+
+    Args:
+        desired_percentiles:
+            The target percentiles.
+        original_percentiles:
+            The percentiles at which the data is provided.
+        data:
+            The data to be interpolated.
+        max_value:
+            The maximum potential value of the percentile coordinate, defaults
+            to 100. Used to calculate an offset for vectorisation.
+
+    Returns:
+        The forecast values interpolated to the desired_percentiles.
+    """
+
+    def calculate_offset(percentiles_size: int, dim_size: int):
+        """Calculate offsets, each step has a length determined by the variable
+        percentile_size. The number of steps is set by dim_size. The increment
+        is set by max_value. An array describing the offsets is returned."""
+        steps = []
+        for i in range(0, dim_size):
+            steps.extend(np.ones(percentiles_size, dtype=np.int) * i * max_value)
+        return np.array(steps).reshape(percentiles_size, dim_size)
+
+    desired_percentiles = np.array(desired_percentiles)
+    loop_size = data.shape[0]
+    # Calculate the remaining array size that is not part of the looping (first)
+    # dimension or the percentile (last) dimension.
+    vectorised_dim_size = data[..., -1].size // loop_size
+
+    # Flatten input and target percentiles and tile
+    orig_flat = (
+        np.tile(original_percentiles, vectorised_dim_size)
+        .reshape(original_percentiles.size, -1)
+        .astype(np.float64)
+    )
+    desired_flat = (
+        np.tile(desired_percentiles, vectorised_dim_size)
+        .reshape(desired_percentiles.size, -1)
+        .astype(np.float64)
+    )
+    # Offset the repeated sets of percentiles to make them distinct
+    orig_flat += calculate_offset(original_percentiles.size, vectorised_dim_size)
+    desired_flat += calculate_offset(desired_percentiles.size, vectorised_dim_size)
+
+    # Interpolate along one entire span of the looping (first) dimension in
+    # each iteration of the loop before reshaping the output.
+    result = []
+    for index in range(loop_size):
+        result.append(
+            np.interp(
+                desired_flat.reshape(-1),
+                orig_flat.reshape(-1),
+                data[index].reshape(-1),
+            ).reshape(vectorised_dim_size, -1)
+        )
+
+    # Construct an array of shape
+    # (loop_size, vectorised_dim_size, desired_percentiles.size)
+    result = np.stack(result).astype(np.float32)
```

Comment:
I think it would be more efficient to pre-allocate ``result``:
```
    result = np.empty((loop_size, vectorised_dim_size, desired_percentiles.size), dtype=np.float32)
    for index in range(loop_size):
        result[index] = np.interp(
                desired_flat.reshape(-1),
                orig_flat.reshape(-1),
                data[index].reshape(-1),
            ).reshape(vectorised_dim_size, -1)
```

Reply:
Done.